File: day2/day2.py
#!/usr/bin/env python3

kp1 = (
    ( "", "", "", "", "" ),
    ( "", "1", "2", "3", "" ),
    ( "", "4", "5", "6", "" ),
    ( "", "7", "8", "9", "" ),
    ( "", "", "", "", "" ),
)

kp2 = (
    ( "", "", "", "", "", "", "" ),
    ( "", "", "", "1", "", "", "", ),
    ( "", "", "2", "3", "4", "", "", "" ),
    ( "", "5", "6", "7", "8", "9", "" ),
    ( "", "", "A", "B", "C", "", "" ),
    ( "", "", "", "D", "", "", "" ),
    ( "", "", "", "", "", "", "" ),
)

def decode(txt, keypad, x, y):
    nums = []
    for row in txt.split("\n"):
        for c in row:
            if c == "U":
                if keypad[y-1][x]:
                    y -= 1
            elif c == "D":
                if keypad[y+1][x]:
                    y += 1
            elif c == "L":
                if keypad[y][x-1]:
                    x -= 1
            elif c == "R":
                if keypad[y][x+1]:
                    x += 1
        nums.append(keypad[y][x])

    return "".join(nums)

# txt ="""ULL
# RRDDD
# LURDL
# UUUUD"""
# assert decode(txt, kp1, 2, 2) == "1985"
# assert decode(txt, kp2, 1, 3) == "5DB3"

import sys
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: {} <filename>".format(sys.argv[0]), file=sys.stderr)
        sys.exit(1)

    try:
        with open(sys.argv[1], "rt") as f:
            txt = f.read().strip()
    except:
        print("Cannot open {}".format(sys.argv[1]), file=sys.stderr)
        sys.exit(1)

    print("Part1:", decode(txt, kp1, 2, 2))
    print("Part2:", decode(txt, kp2, 1, 3))

File: day2/day2.cpp
#include <fstream>
#include <fmt/format.h>

using namespace std;

static const char keypad1[][7] = {
	{ 0,   0,   0,   0, 0, 0, 0 },
	{ 0, '1', '2', '3', 0, 0, 0 },
	{ 0, '4', '5', '6', 0, 0, 0 },
	{ 0, '7', '8', '9', 0, 0, 0 },
	{ 0,   0,   0,   0, 0, 0, 0 },
};

static const char keypad2[][7] = {
	{ 0,   0,   0,   0,   0,   0, 0 },
	{ 0,   0,   0, '1',   0,   0, 0 },
	{ 0,   0, '2', '3', '4',   0, 0 },
	{ 0, '5', '6', '7', '8', '9', 0 },
	{ 0,   0, 'A', 'B', 'C',   0, 0 },
	{ 0,   0,   0, 'D',   0,   0, 0 },
	{ 0,   0,   0,   0,   0,   0, 0 },
};

static string decode(const vector<string>& instr, const char keypad[][7], int x, int y)
{
	string rv;
	for (auto& s: instr)
	{
		for (auto c: s)
		{
			int nx = x;
			int ny = y;
			switch (c)
			{
			case 'U': ny--;	break;
			case 'L': nx--; break;
			case 'R': nx++; break;
			case 'D': ny++; break;
			default:
				continue;
			}
			if (keypad[ny][nx])
			{
				x = nx;
				y = ny;
			}
		}
		rv.push_back(keypad[y][x]);
	}
	return rv;
}

int main(int argc, char *argv[])
{
	if (argc < 2)
	{
		fmt::print(stderr, "Usage: {} <filename>\n", argv[0]);
		return 1;
	}

	ifstream input(argv[1]);
	if (!input)
	{
		fmt::print(stderr, "Cannot open {}\n", argv[1]);
		return 1;
	}

	vector<string> lines;
	string line;
	while (getline(input, line))
	{
		lines.emplace_back(move(line));
	}
	input.close();

	fmt::print(stderr, "Part1: {}\n", decode(lines, keypad1, 2, 2));
	fmt::print(stderr, "Part2: {}\n", decode(lines, keypad2, 1, 3));
	return 0;
}
